pkg/ipam/state.go

```diff
@@ -0,0 +1,417 @@
+// Copyright 2016-2019 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ipam
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net"
+	"os"
+	"path"
+	"sort"
+	"strings"
+
+	"github.com/cilium/cilium/pkg/k8s"
+)
+
+// This submodule extends the cilium IPAM by providing a persistent storage
+// for storing IP states.
+// The main purpose is to support fixed IP pods for statefulsets.
+
+// EnableFixedIP flag indicates whether fixed IP is enabled
+var EnableFixedIP bool = true
+
+const (
+	enableFixedIPEnv string = "ENABLE_FIXED_IP"
+	ipamStateDir     string = "/var/lib/ipamstate"
+	minFileNameLen   int    = 9 // format: <ns>-<pod>:<ip>
+)
+
+// cleanStaleIPPinnings cleans the stale pinned IP states.
+// This function should be called on agent start/restart
+func cleanStaleIPPinnings(ipam *IPAM) error {
+	pinnedIPs := []string{}
+	pinnedOwners := []string{}
+	log.Infof("Cleaning stale IP pinnings")
+
+	nodeName := os.Getenv(k8s.EnvNodeNameSpec)
+	if nodeName == "" {
+		err := fmt.Errorf("Get %s from ENV failed", k8s.EnvNodeNameSpec)
+		log.Errorf("Clean stale IP pinnings failed: %s, skip cleaning", err)
+		return err
+	}
+
+	podNames, err := k8s.GetNodeStsPods(nodeName)
+	if err != nil {
+		log.Errorf("Get node sts pod list failed: %s, skip cleaning", err)
+		return err
+	}
+
+	log.Infof("sts pods should on this node: %s", podNames)
+
+	files, err := ioutil.ReadDir(ipamStateDir)
+	if err != nil {
+		log.Errorf("Read %s failed: %s, skip cleaning", ipamStateDir, err)
+		return err
+	}
+
+	sort.Strings(podNames)
+	for _, f := range files {
+		name := f.Name()
+		if len(name) < minFileNameLen {
+			log.Errorf("Clean stale IP pinnings: unknown pin info %s, skip "+
+				"this one", name)
+			continue
+		}
+
+		podName := ""
+		for _, pname := range podNames {
+			// TODO: break early as we are sorted slice
+
+			if strings.HasPrefix(name, strings.ReplaceAll(pname, "/", "-")) {
+				podName = pname
+				break
+			}
+		}
+
+		if podName != "" {
+			items := strings.Split(name, ":")
+			if len(items) != 2 {
+				log.Errorf("Unknown pinning info: %s, skip this one", name)
+				continue
+			}
+
+			// Note: better to read the owner name from file content
+			// replace '-' with '/' is 100% correct for sts?
+			pinnedOwners = append(pinnedOwners, strings.ReplaceAll(items[0], "-", "/"))
+			pinnedIPs = append(pinnedIPs, items[1])
+			log.Infof("Pod %s exists in k8s, keep file %s", podName, name)
+			continue
+		}
```

Comment:
This needs to live in a function with unit tests made

Reply:
Done.